app/lib/search/updater.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:gcloud/db.dart';
+
+import '../frontend/models.dart';
+import '../shared/mock_scores.dart';
+import '../shared/search_service.dart';
+import '../shared/utils.dart' show sliceList;
+
+Logger _logger = new Logger('pub.search.updater');
+
+class PackageIndexUpdater {
+  final PackageIndex index;
+  Timer _timer;
+  bool _updating = false;
+
+  PackageIndexUpdater(this.index);
+
+  Future update({int limit}) async {
+    if (_updating) {
+      _logger.warning('Update called before previous round completed.');
+      return;
+    }
+    try {
+      _updating = true;
+      await _doUpdate(limit);
+    } catch (e, st) {
+      _logger.severe('Error while updating search index.', e, st);
+    } finally {
+      _updating = false;
+    }
+  }
+
+  Future _doUpdate(int limit) async {
+    final List<Key> versionKeys = [];
+    final Map<String, String> devVersions = {};
+
+    int count = 0;
+    await for (Package p in dbService.query(Package).run()) {
+      versionKeys.add(p.latestVersionKey);
+      devVersions[p.name] = p.latestDevVersion;
+      count++;
+      if (limit != null && limit <= count) break;
+    }
+
+    for (List<Key> slices in sliceList(versionKeys, 100)) {
```

Comment:
Maybe lower the list from 100 to more like 20 here.

The issue is that `PackageVersion` contains README/CHANGELOG/pubspec.yaml so they can make the protobuf we get back quite big. There is a limit on how big a response to a single lookup request can be, so to avoid flakiness it might make sense to lower this from the current 100.